### PR TITLE
Tag

### PR DESCRIPTION
**Pull request by MontrealSergiy**
tag/rev info refactored with git tag describe

### CODE REVIEW

**Comment by MontrealSergiy** [comment]
found bug

**Comment by prioux** [comment]
I was about to comment.

I will still do it in the code section here even if it's closed.

File: BrainPortal/lib/cbrain_file_revision.rb [hunk 1]
```diff
@@ -409,32 +409,16 @@ def self.cbrain_head_tag(what = 'CBRAIN') #:nodoc:
...
-        git_tag += "-#{num_new_commits}" if num_new_commits > 0
-      end
-    #end
+    cd_bash = "cd #{what_root.to_s.bash_escape} ; " # Prefix for all bash commands below
+
+    return DEFAULT_TAG if `#{cd_bash} git rev-parse --show-toplevel 2>/dev/null`.strip != what_root.to_s
 
-    git_tag # not cached, so it's really live
+    #Dir.chdir(what_root.to_s) do
+      out = `git describe --tags --long`
```
Comment (prioux) [inline BrainPortal/lib/cbrain_file_revision.rb:420]: 1) Indentation.

2) you forgot to do the `cd` command so this code no longer extracts the revision numbers for the plugins.

Comment (prioux) [reply BrainPortal/lib/cbrain_file_revision.rb:420]: 2) many of the variables defined above no longer are useful (seen, what_root etc), so clean up

Comment (prioux) [reply BrainPortal/lib/cbrain_file_revision.rb:420]: 4) Don't leave the commented-out Dir.chdir statement, it's useless now.

File: BrainPortal/script/show_cbrain_rev [hunk 2]
```diff
@@ -34,25 +34,10 @@ end
...
-        num_new_commits = `git rev-list '#{git_tag}..HEAD'`.split.size
-      elsif default_tag # fallback for plugins, when -z used
-        num_new_commits = `git rev-list HEAD`.split.size
-        git_tag = default_tag
-      end
-      if git_tag
-        git_tag += "-#{num_new_commits}" if num_new_commits > 0
+      out = `git describe --tags --long`
+      if out
+        git_tag, _, _ = out.rpartition('-')
```
Comment (prioux) [inline BrainPortal/script/show_cbrain_rev:40]: why do you use this strange statement when in the other code you used `out.rpartition('-')[0]` ?

Comment (prioux) [reply BrainPortal/script/show_cbrain_rev:40]: to be clean, redirect stderr in the git command

Comment (MontrealSergiy) [reply BrainPortal/script/show_cbrain_rev:40]: ok for '-'. that is a common pattern with python, so I used with it.

**Comment by MontrealSergiy** [comment]
@prioux Are we supporting ruby 2.4 ? (in 2.5 offers one convenient sting method)

File: BrainPortal/script/show_cbrain_rev (hunk 2 above)
Comment (MontrealSergiy) [reply BrainPortal/script/show_cbrain_rev]: sorry not follow sderr to std out ? or null?

File: BrainPortal/lib/cbrain_file_revision.rb (hunk 1 above)
Comment (MontrealSergiy) [reply BrainPortal/lib/cbrain_file_revision.rb:420]: `what_root` is used for cd (which you asked to put back)

File: BrainPortal/script/show_cbrain_rev (hunk 2 above)
Comment (prioux) [reply BrainPortal/script/show_cbrain_rev]: To /dev/null . Redirect the command's stderr to /dev/null so we don't pollute the output with spurious error messages. **Just like everywhere else in this file!**

Comment (MontrealSergiy) [reply BrainPortal/script/show_cbrain_rev]: adding command to that file (not sure for another)

File: BrainPortal/lib/cbrain_file_revision.rb [hunk 3]
```diff
@@ -403,39 +403,18 @@ def self.cbrain_head_tag(what = 'CBRAIN') #:nodoc:
...
-      if git_tag
-        git_tag += "-#{num_new_commits}" if num_new_commits > 0
-      end
-    #end
-
-    git_tag # not cached, so it's really live
+    return DEFAULT_TAG if `#{cd_bash} git rev-parse --show-toplevel 2>/dev/null`.strip != what_root.to_s
 
+    out = `#{cd_bash} git describe --tags --long 2>/dev/null`
+    if out.present?
```
Comment (prioux) [inline BrainPortal/lib/cbrain_file_revision.rb:414]: I won't ask you to change it but just so you know, you could have done all four lines of your `if` block with a single line instead

```
out.gsub!(/-0$/,"")
```

You don't even have to verify it ends with -0 or not: if the substitute fails, it leaves the string alone.

File: BrainPortal/script/show_cbrain_rev [hunk 4]
```diff
@@ -34,26 +34,11 @@ end
...
-      if git_tag
-        num_new_commits = `git rev-list '#{git_tag}..HEAD'`.split.size
-      elsif default_tag # fallback for plugins, when -z used
-        num_new_commits = `git rev-list HEAD`.split.size
-        git_tag = default_tag
-      end
-      if git_tag
-        git_tag += "-#{num_new_commits}" if num_new_commits > 0
-        puts git_tag
+      out = `git describe --tags --long`
```
Comment (prioux) [inline BrainPortal/script/show_cbrain_rev:38]: Same here.. Al of this could be

```
  `git describe --tags --long`
  .strip
  .sub(/(-0)?-\w+$/,"")
```

Comment (MontrealSergiy) [reply BrainPortal/script/show_cbrain_rev:38]: thanks, but can replace if you prefer. I know that regex is powerful and concise, yet I believe tat re is a bit condense for most people, me included, so we have more trouble with deciphering regexes than with string operations involving partition/delete_suffix/split and alike, so tried ditch re.

Comment (prioux) [reply BrainPortal/script/show_cbrain_rev:38]: That's a terrible argument. This regex is very simple. If you think it's still too complex then you write it with the extended syntax that allows you to embed comments:

```
  `git describe --tags --long`
  .strip
  .sub(
    /
      (-0)?           # optional dash zero, like "-0"
      -\w+            # a dash and alphanums like "-b027af3"
     $                  # anchor at end of string
    /x ,
    ""                  # replace with nothing
 )
```

To me that is overkill, and the straight version ` .sub( /(-0)?-\w+$/,  "" ) ` is totally readable.

Comment (MontrealSergiy) [reply BrainPortal/script/show_cbrain_rev:38]: well, IMHO. stackoverflow is filled with regex hacks question. I shortened the old version a bit, hope better now

**Comment by prioux** [comment]
Your code still doesn't work properly. I said I wanted no change in behavior.

I'm running your script in a plugin with no tags I a get a STDERR message.

Here's what my original code did:

```
> cd /BrainPortal/cbrain_plugins/cbrain-plugins-bids-fmriprep # any plugins with no tags
> ../../script/show_cbrain_rev -z
0.1.0-14
```

Here's what your code does:

```
> ../../script/show_cbrain_rev -z
fatal: No names found, cannot describe anything.

```

(one error line is on stderr, and one blank line on stdout)

**Review by prioux — CHANGES_REQUESTED** [review]
Still buggy. See my last comment.

Also remember to redirect stderr in the mail CBRAIN code too.

**Comment by MontrealSergiy** [comment]
sorry, I think you meant to redirect stderr in standalone secondary script?

**Comment by MontrealSergiy** [comment]
done

**Pull request merged**